src/test/java/com/iota/iri/service/tipselection/impl/WalkValidatorImplTest.java

```diff
@@ -251,9 +241,102 @@ public void failOnInconsistency() throws Exception {
-                milestoneTracker, 15, Integer.parseInt(Configuration.BELOW_MAX_DEPTH_LIMIT),
-                Integer.parseInt(Configuration.WALK_VALIDATOR_CACHE));
+                milestoneTracker, 15, Integer.parseInt(Configuration.BELOW_MAX_DEPTH_LIMIT));
         Assert.assertFalse("Validation succeded but should have failed due to inconsistent ledger state",
                 walkValidator.isValid(hash));
     }
+
+    @Test
+    public void dontMarkWrongTxsAsBelowMaxDepth() throws Exception {
+        final int maxAnalyzedTxs = Integer.parseInt(Configuration.BELOW_MAX_DEPTH_LIMIT);
+        TransactionViewModel tx1 = TransactionTestUtils.createBundleHead(0);
+        tx1.store(tangle);
+        tx1.setSnapshot(tangle, 92);
+
+        TransactionViewModel txBad = TransactionTestUtils.createBundleHead(0);
+        txBad.store(tangle);
+        txBad.setSnapshot(tangle, 10);
+
+        TransactionViewModel tx2 = new TransactionViewModel(TransactionViewModelTest.getRandomTransactionWithTrunkAndBranch(tx1.getHash(), tx1.getHash()),
+                TransactionViewModelTest.getRandomTransactionHash());
+        TransactionTestUtils.setLastIndex(tx2,0);
+        TransactionTestUtils.setCurrentIndex(tx2,0);
+        tx2.updateSolid(true);
+        tx2.store(tangle);
+
+        TransactionViewModel tx3 = new TransactionViewModel(TransactionViewModelTest.getRandomTransactionWithTrunkAndBranch(tx1.getHash(), txBad.getHash()),
+                TransactionViewModelTest.getRandomTransactionHash());
+        TransactionTestUtils.setLastIndex(tx3,0);
+        TransactionTestUtils.setCurrentIndex(tx3,0);
+        tx3.updateSolid(true);
+        tx3.store(tangle);
+
+        TransactionViewModel tx4 = new TransactionViewModel(TransactionViewModelTest.getRandomTransactionWithTrunkAndBranch(tx2.getHash(), tx3.getHash()),
+                TransactionViewModelTest.getRandomTransactionHash());
+        TransactionTestUtils.setLastIndex(tx4,0);
+        TransactionTestUtils.setCurrentIndex(tx4,0);
+        tx4.updateSolid(true);
+        tx4.store(tangle);
+
+        Mockito.when(ledgerValidator.updateDiff(new HashSet<>(), new HashMap<>(), tx4.getHash()))
+                .thenReturn(true);
+        Mockito.when(ledgerValidator.updateDiff(new HashSet<>(), new HashMap<>(), tx2.getHash()))
+                .thenReturn(true);
+
+        milestoneTracker.latestSolidSubtangleMilestoneIndex = 100;
+        WalkValidatorImpl walkValidator = new WalkValidatorImpl(tangle, ledgerValidator, transactionValidator,
+                milestoneTracker, 15, maxAnalyzedTxs);
+        Assert.assertFalse("Validation of tx4 succeeded but should have failed since tx is below max depth",
+                walkValidator.isValid(tx4.getHash()));
+        Assert.assertTrue("Validation of tx2 failed but should have succeeded since tx is above max depth",
+                walkValidator.isValid(tx2.getHash()));
```

Comment:
From Tx4 to reach the TxBad you need to walk branch, branch (2 steps)
From Tx2 to reach Tx1 (good) you need to walk branch or trunk (1 steps)

IMO you should construct Tx2 as close as possible to Tx4.